### PR TITLE
ConcatSeqsDataset pad_narrow_data_to_multiple_of_target_len

### PR DESCRIPTION
**Pull request by Stefanwuu**
To allow training with alignments of raw wav data, we make sure every concatenated sequence has strictly aligned length

### CODE REVIEW

File: returnn/datasets/meta.py [hunk 1]
```diff
@@ -1390,7 +1390,7 @@ def __init__(
         seq_len_file,
         seq_tag_delim=";",
         remove_in_between_postfix=None,
-        repeat_in_between_last_frame_up_to_multiple_of=None,
+        force_align=None,
```
Comment (NeoLegends) [inline returnn/datasets/meta.py:1393]: Renaming/changing the purpose of this parameter is going to break existing setups that use this dataset. Usually we try to stay backwards compatible.

One solution here is to keep both parameters (and their defaults), and then assert that only one of them is set at a time and then apply the corresponding logic.

I'm also not sure `force_align` is a good name for this key, as AFAIK the term "forced alignment" is usually used for producing an alignment from a trained model. I'd rather go with something like `force_pad_truncate_data_to_target_length`.

File: returnn/datasets/meta.py [hunk 2]
```diff
@@ -1557,11 +1555,17 @@ def _collect_single_seq(self, seq_idx):
...
+                if key in self.force_align and sub_seq_idx != sub_seq_idxs[-1]:
+                    ref_key, frame_rate = self.force_align[key]
+                    ref_data = self.sub_dataset.get_data(sub_seq_idx, ref_key)
+                    len_diff = data.shape[0] - ref_data.shape[0] * frame_rate
+                    if len_diff > 0:
+                        # if data longer than ref_data * frame_rate, cut down
+                        data = data[:-len_diff,]
+                    elif len_diff < 0:
+                        # if data shorter than ref_data * frame_rate, pad by repeating last frame
+                        data = numpy.concatenate([data] + [data[-1:]] * -len_diff, axis=0)
```
Comment (NeoLegends) [inline returnn/datasets/meta.py:1567]: Should there be a sanity limit how much can be truncated or padded? Perhaps to allow padding/truncating only within one multiple of the target length vs. feature length? In my own experiments w/ subsampled MFCCs the mismatch was generally within this range (but then it was also MFCCs and not raw WAV...).

**Review by NeoLegends — CHANGES_REQUESTED** [review]
Thank you for filing this PR! I believe there are backwards incompatible changes which should probably be implemented differently.

File: returnn/datasets/meta.py (hunk 2 above)
Comment (michelwi) [reply returnn/datasets/meta.py:1567]: Compared to features, for raw wav the mismatch will generally be larger by a factor of sampling_rate * frame_time e.g. 160 for 16kHz audio and 10ms frames.

File: returnn/datasets/meta.py (hunk 1 above)
Comment (Stefanwuu) [reply returnn/datasets/meta.py:1393]: Thank you for your information, I did this because (to my best knowledge) nobody used this in training before:) But you are right, such potential issue should be avoided, I will add back the parameter before and change the name

File: returnn/datasets/meta.py (hunk 2 above)
Comment (Stefanwuu) [reply returnn/datasets/meta.py:1567]: Do you mean sth like `assert len_diff <= frame_rate`? I think there is no clear upper bound if raw WAV is used. The mismatch is partially setup dependent(window_size - window_step in feature extraction) and can be more than one frame(10 ms). Maybe it would make sense to throw a warning instead of a hard limit?

**Comment by albertz** [comment]
So this PR here is to fix issue #1573?

File: returnn/datasets/meta.py [hunk 3]
```diff
@@ -1486,7 +1491,7 @@ def get_seq_len(i):
             sub_seq_list.extend(sub_seq_tags)
         assert sub_seq_idx == len(sub_seq_list) and len(seq_list) == len(sub_seq_idxs)
         self.cur_sub_seq_idxs = sub_seq_idxs
-        return self.sub_dataset.init_seq_order(seq_list=sub_seq_list)
+        return self.sub_dataset.init_seq_order(epoch=epoch, seq_list=sub_seq_list)
```
Comment (albertz) [inline returnn/datasets/meta.py:1494]: I wonder a bit, why this change? It should be ok, but actually also should not matter? Or was there a problem? I'm just curious.

File: returnn/datasets/meta.py [hunk 4]
```diff
@@ -1539,6 +1544,14 @@ def _collect_single_seq(self, seq_idx):
...
                     sub_dataset_keys,
                 )
+            for key in self.pad_truncate_data_to_multiple_of:
+                assert (
+                    key in sub_dataset_keys
+                ), "%s: pad_truncate_data_to_multiple_of key %r not in sub dataset data-keys %r" % (
+                    self,
+                    key,
+                    sub_dataset_keys,
+                )
```
Comment (albertz) [inline returnn/datasets/meta.py:1554]: Better use f-strings.
```suggestion
                assert key in sub_dataset_keys, (
                    f"{self}: pad_truncate_data_to_multiple_of key {key}"
                    f" not in sub dataset data-keys {sub_dataset_keys}"
                )
```

File: returnn/datasets/meta.py [hunk 5]
```diff
@@ -1606,6 +1630,12 @@ def get_data_shape(self, key):
         """
         return self.sub_dataset.get_data_shape(key)
 
+    def get_total_num_seqs(self):
+        """
+        :rtype: int
+        """
```
Comment (albertz) [inline returnn/datasets/meta.py:1636]:
```suggestion
    def get_total_num_seqs(self) -> int:
        """total num seqs"""
```

File: returnn/datasets/meta.py (hunk 1 above)
Comment (albertz) [reply returnn/datasets/meta.py]: Note that I know that people did use `ConcatSeqsDataset` in training, even recently. :) (Although not sure if this specific option was used.)

**Comment by albertz** [comment]
Failing tests?

File: returnn/datasets/meta.py [hunk 6]
```diff
@@ -1406,13 +1407,17 @@ def __init__(
           Now it could happen that ceildiv(data_len1 + data_len2, 6) < align_len1 + align_len2.
           This option would repeat intermediate ending frames such that data_len1 % 6 == 0,
           by setting it to {"data": 6}.
+        :param dict[str,(str,int)]|None pad_truncate_data_to_multiple_of: data_key -> (target_key, multiple).
+            to make sure the alignment is still valid after concatenation, we force length of every seq in data_key
+            to be length of seq in target_key * multiple
```
Comment (albertz) [inline returnn/datasets/meta.py:1412]: I find the documentation a bit misleading. Specifically "to make sure the alignment is still valid after concatenation". Specifically, the option `repeat_in_between_last_frame_up_to_multiple_of` is also there to handle this issue already. So you should better explain, why is this new option needed. What would be a situation where `repeat_in_between_last_frame_up_to_multiple_of` would not work.

**Comment by albertz** [comment]
On the method itself: The first question is, how is the actual relation between input len and output len?

Note e.g. in `repeat_in_between_last_frame_up_to_multiple_of`, I mention the relation $\lceil T_{\textrm{in}} / F \rceil = T_{\textrm{out}}$ where $F$ is the downsampling rate. This is what you get when you have padding=same.

For padding=valid (i.e. no padding allowed for the window, the window must always cover real frames), you would have $\lceil ( T_{\textrm{in}} - (W - 1) ) / F \rceil = T_{\textrm{out}}$, where $W$ is the window/kernel size.

With feature extraction on raw wav, we have actually various different variants. Most implementations in TF and/or PT would use `stft` in some way, which usually is padding=valid.

Note that any custom padding schemes are possible.

Padding=same is actually just designed in a way to get this nice $\lceil T_{\textrm{in}} / F \rceil = T_{\textrm{out}}$ relation (or with $F=1$ you simply have $T_{\textrm{in}} = T_{\textrm{out}}$, thus "same"). Padding=same means it will pad $W - 1$ frames, evenly distributed left/right (right one more if uneven).

The `padding` option in Torch (`Conv` and co) specifies usually how many frames are padded left and right (so it means for `padding=P`, it will actually pad 2*P frames in total).

RASR feature extraction, for some strange reasons, does pad some totally custom number of frames: It pads in total $F-1$ frames. I.e. you have the relation $\lceil ( T_{\textrm{in}} - (W - 1) + (F - 1)) / F \rceil = \lceil ( T_{\textrm{in}} - W + F) / F \rceil = T_{\textrm{out}}$.

Btw, also see https://github.com/rwth-i6/returnn/issues/1564 for a recent discussion on exactly the same topic.

I already commented on improving the documentation. I think these different cases should become clear in the documentation.

`repeat_in_between_last_frame_up_to_multiple_of` is right when you have padding=same.

`pad_truncate_data_to_multiple_of` is actually also only right when you have padding=same, or not?

File: returnn/datasets/meta.py (hunk 3 above)
Comment (Stefanwuu) [reply returnn/datasets/meta.py:1494]: The problem is it could be incompatible with some sub_dataset like `MetaDataset`, where init_seq_order would not be actually executed if `epoch` stays `None`, maybe there are other workarounds?

Comment (albertz) [reply returnn/datasets/meta.py:1494]: Ah yea, good example. Yes this is definitely more correct.

File: returnn/datasets/meta.py [hunk 7]
```diff
@@ -1406,13 +1407,19 @@ def __init__(
           Now it could happen that ceildiv(data_len1 + data_len2, 6) < align_len1 + align_len2.
           This option would repeat intermediate ending frames such that data_len1 % 6 == 0,
           by setting it to {"data": 6}.
+        :param dict[str,(str,int)]|None pad_truncate_data_to_multiple_of: data_key -> (target_key, multiple).
+            Force data_len == target_len * multiple for specified data and target.
+            Similar to repeat_in_between_last_frame_up_to_multiple_of but also truncates the data.
```
Comment (albertz) [inline returnn/datasets/meta.py:1412]: > Similar to repeat_in_between_last_frame_up_to_multiple_of but also truncates the data.

This does not really explain it. In what case would it be relevant to truncate the data?

File: returnn/datasets/meta.py [hunk 8]
```diff
@@ -1406,13 +1407,19 @@ def __init__(
...
+            Mainly for raw wav alignment, where data_len could be slightly longer than target_len * multiple
+            due to mismatch in feature extraction.
```
Comment (albertz) [inline returnn/datasets/meta.py:1414]: "raw wav alignment" doesn't really say anything / is not relevant here.

What is the case you have specifically, i.e. what type of padding was done? This is what is relevant.

Why is `repeat_in_between_last_frame_up_to_multiple_of` not already enough?

File: returnn/datasets/meta.py (hunk 7 above)
Comment (Stefanwuu) [reply returnn/datasets/meta.py:1412]: I know, this is not the final version as mentioned in the commit comment. I want to first consider your long comment above that I haven't read through, maybe there is indeed situations(especially with padding) that is not covered by my code.

File: returnn/datasets/meta.py [hunk 9]
```diff
@@ -1391,6 +1391,7 @@ def __init__(
         seq_tag_delim=";",
         remove_in_between_postfix=None,
         repeat_in_between_last_frame_up_to_multiple_of=None,
+        pad_truncate_data_to_multiple_of=None,
```
Comment (albertz) [inline returnn/datasets/meta.py:1394]: Shouldn't there also be `target_len` or so in the name?

File: returnn/datasets/meta.py (hunk 8 above)
Comment (Stefanwuu) [reply returnn/datasets/meta.py:1414]: The main problem I want to tackle here is indeed relevant to "raw wav alignment"(but of course also to the padding scheme). The main problem with raw wav is that even with same padding, we could have $\left \lceil{T_{in}/F}\right \rceil \neq T_{out}$. With most common `stft` setting, we have `frame_size` of 25ms and `frame_shift` of 10 ms. This means a perfectly aligned 3-frame text corresponds to 45ms audio data, and you can see $\left \lceil{45/10}\right \rceil \neq 3$. In chunked training, we solve this by setting min_chunk_size to a magic number(which is not elegant, but we all do this, see [more explanation](https://omnifluent.atlassian.net/wiki/spaces/AS/pages/2636709904/Deep+dive+into+NN+training#Chunking)). But this gets wrong if such mismatch adds up with sequence concatenation. That's why I truncate such mismatches for all sub-sequences except the last one.

File: returnn/datasets/meta.py [hunk 10]
```diff
@@ -1406,13 +1407,19 @@ def __init__(
...
+            Mainly for raw wav alignment, where data_len could be slightly longer than target_len * multiple
+            due to mismatch in feature extraction.
         :param bool use_cache_manager:
         :param dict[(int,int),dict] epoch_wise_filter: see :class:`EpochWiseFilter`
         """
         super(ConcatSeqsDataset, self).__init__(**kwargs)
         self.seq_tag_delim = seq_tag_delim
         self.remove_in_between_postfix = remove_in_between_postfix or {}
         self.repeat_in_between_last_frame_up_to_multiple_of = repeat_in_between_last_frame_up_to_multiple_of or {}
+        self.pad_truncate_data_to_multiple_of_target_len = pad_truncate_data_to_multiple_of_target_len or {}
```
Comment (NeoLegends) [inline returnn/datasets/meta.py:1422]: I think this should have an assertion that not both properties are set?

File: returnn/datasets/meta.py (hunk 8 above)
Comment (albertz) [reply returnn/datasets/meta.py:1414]: > The main problem I want to tackle here is indeed relevant to "raw wav alignment"(but of course also to the padding scheme).

This doesn't make sense. It can not be relevant that you operate on raw wav. It only matters what kind of downsampling/padding scheme you have. Why don't you say what case you have actually? What code are you using? You can simply look that up.

> The main problem with raw wav is that even with same padding, we could have $\left \lceil{T_{in}/F}\right \rceil \neq T_{out}$

No. Padding=same implies that you have $\left \lceil{T_{\textrm{in}}/F}\right \rceil = T_{\textrm{out}}$, always. That follows directly from the definition of padding=same. That's actually what it means. If you don't have $\left \lceil{T_{\textrm{in}}/F}\right \rceil = T_{\textrm{out}}$, then it means, you don't have padding=same but you have some different padding scheme.

So I get back to my question: What padding scheme do you have?

> With most common `stft` setting, ...

I assume you have padding=valid then?

> In chunked training, we solve this by setting min_chunk_size to a magic number

Was this related to the discussion here? https://github.com/rwth-i6/returnn/issues/1564

But again, the main question: What type of padding do you use? Or if you don't know and cannot tell from your code, can you at least say what code you use?

Comment (Stefanwuu) [reply returnn/datasets/meta.py:1414]: Hmm, I thought I knew which padding you were talking about but now I'm not sure. Do you refer to the padding/downsampling scheme in RETURNN? E.g. `pad_last_frame` in `MultiChannelStftLayer` or same padding in conv layer if conformer was used for downsampling?

Comment (albertz) [reply returnn/datasets/meta.py:1414]: No, I mean whatever you did to get your original input len <-> output len relation. I don't care about what software you used. I just want to know your exact input len <-> output len relation, i.e. the type of padding scheme you have used there.

Comment (Stefanwuu) [reply returnn/datasets/meta.py:1414]: But that's why raw wav is relevant here. In raw wav, the first frame is longer than the rest of frames due to mismatched window size and window shift step in feature extraction(see my example above). A padding can't fix this because it's a problem of too long data instead of too short data. That's why I propose to truncate in ConcatSeqsDataset in addition to the padding you implemented before.

Comment (Stefanwuu) [reply returnn/datasets/meta.py:1414]: Sorry for the confusion, I had a bit discussion with Moritz and it turns out I didn't understand your point on padding. 
The main issue here is the padding logic in my RETURNN in-network feature extraction, as you described
> For padding=valid (i.e. no padding allowed for the window, the window must always cover real frames), you would have ⌈(Tin−(W−1))/F⌉=Tout, where W is the window/kernel size.

does not fit the padding logic in RASR alignment generation. So yes, it is possible to modify the alignment somewhere else to avoid the modification of RETURNN. Then maybe I should close this pull request?

Comment (albertz) [reply returnn/datasets/meta.py:1414]: I still don't understand your padding scheme. What *exactly* is your input len <-> output len relation of the given data you have? There must be an *exact* formula to describe that relation. If you don't know the formula, please find out. It would also help when you say what code in your case determines this relation. E.g. what is your feature extraction? I didn't really understand whether you use RETURNN, RASR, or sth else to generate this data, and in any case, what exactly.

Whatever the formula is, from the formula, you can infer your padding scheme. Nothing else is relevant. It's still not relevant that you operate on raw wav.

Comment (Stefanwuu) [reply returnn/datasets/meta.py:1414]: Yes, I understand your point that operating on raw WAV is not relevant now, only the alignment logic applied there is relevant. Sorry again for the misunderstanding before. The exact formula in my setup(slightly different from most common setup): 
$T_{out}=\left \lfloor{(T_{in}-(W-F))/F} \right \rfloor$
, which is different from the alignment(or padding) scheme of RASR-generated alignment.

Comment (albertz) [reply returnn/datasets/meta.py:1414]: Really floordiv, not ceildiv? Can you give a reference to the code which leads to such formula?

Comment (albertz) [reply returnn/datasets/meta.py:1414]: Assuming the formula is correct: You can translate this into the standard scheme of the formula:
$T_{out} = \lfloor (T_{in}-(W-F))/F \rfloor = \lceil (T_{in} - (W - F) - (F - 1)) / F \rceil = \lceil (T_{in} - (W - 1)) / F \rceil$
I.e. it means you have exactly the padding=valid case. That's what I wanted to know.

Comment (Stefanwuu) [reply returnn/datasets/meta.py:1414]: Yes, padding=valid is the case I have. Then what would be your suggestion in this case? Set up the alignment so that padding=valid also holds there instead of modifying RETURNN? Thanks!

**Comment by albertz** [comment]
After some clarification, as I understand you know, I think this new option is intended for the padding=valid case. At least that is your current use case.

Did you actually test the code? Check also my earlier comment here: https://github.com/rwth-i6/returnn/pull/1574#issuecomment-2209637554

I assume you have F=160 (10ms) and W=400 (25ms). And you have padding=valid.

Some example sequences you can have then:
* $T_{in} = 399 \rightarrow T_{out} = 0$
* $T_{in} = 400 \rightarrow T_{out} = 1$
* $T_{in} = 559 \rightarrow T_{out} = 1$
* $T_{in} = 560 \rightarrow T_{out} = 2$
* $T_{in} = 719 \rightarrow T_{out} = 2$
* $T_{in} = 720 \rightarrow T_{out} = 3$

Now when you set `pad_truncate_data_to_multiple_of = {"data": ("targets", 160)`}:

Assume you concat two seqs with $T_{in} = 400, T_{out} = 1$: The first one will be truncated to 160 frames. You get a sequence with $T_{in} = 560, T_{out}=2$. Ok.

Ah, I see now, this should always be ok, no matter the padding. Assume you have a padding scheme like this:
$T_{out} = \lceil (T_{in} - (W + 1) + P) / F \rceil$ for any amount of padded frames $P$. You have $P=0$ (padding=valid).
Now, if you concatenate other seqs, but with your scheme, you make sure that $\Delta T_{out} \cdot F = \Delta T_{in}$ for all but the right outermost seq, so you get for the concatenated seq: $T_{out} + \Delta T_{out} = \lceil (T_{in} + \Delta T_{in} - (W+1) + P) / F \rceil$.

File: returnn/datasets/meta.py (hunk 8 above)
Comment (albertz) [reply returnn/datasets/meta.py:1414]: No. I was simply saying, you should clarify the documentation: you should make clear, for what cases the user would want to use `pad_truncate_data_to_multiple_of`. Why is `repeat_in_between_last_frame_up_to_multiple_of` not enough?

See my last comment in the main thread. The answer is basically: `repeat_in_between_last_frame_up_to_multiple_of` only works for padding=same. But your option works for any padding scheme, e.g. for padding=valid.

Additionally, as you misunderstood for a while this concept of padding scheme (no matter what type of input/output, no matter the software, the input len <-> output len relation can always be expressed as a formula, and that always corresponds to a specific padding scheme), it probably would be helpful to add another sentence or so explaining this.

File: returnn/datasets/meta.py [hunk 11]
```diff
@@ -1391,6 +1391,7 @@ def __init__(
         seq_tag_delim=";",
         remove_in_between_postfix=None,
         repeat_in_between_last_frame_up_to_multiple_of=None,
+        pad_truncate_data_to_multiple_of_target_len=None,
```
Comment (albertz) [inline returnn/datasets/meta.py:1394]: I still don't like the name. `truncate` is misleading here. `truncate` usually means sth different (e.g. see [`torch.trunc`](https://pytorch.org/docs/stable/generated/torch.trunc.html)).

In RETURNN, we have the function `slice_pad_zeros` which does what you want here. So we use the term `slice`. Instead of `slice`, you might also want to call this `narrow` ([`torch.narrow`](https://pytorch.org/docs/stable/generated/torch.narrow.html#torch.narrow)).

So maybe:
```suggestion
        pad_narrow_data_to_multiple_of_target_len=None,
```

**Comment by Stefanwuu** [comment]
> Assume you have a padding scheme like this: Tout=⌈(Tin−(W+1)+P)/F⌉ for any amount of padded frames P. You have P=0 (padding=valid). Now, if you concatenate other seqs, but with your scheme, you make sure that ΔTout⋅F=ΔTin for all but the right outermost seq, so you get for the concatenated seq: Tout+ΔTout=⌈(Tin+ΔTin−(W+1)+P)/F⌉.

Yes, that was precisely my motivation to do this. Sorry for wasting time on the bad communication before, I always considered padding as a specific operation instead of a general formula of alignment. I will update the comment and do some last tests.

**Review by albertz — APPROVED** [review]
See remaining comment by @NeoLegends. Also let's wait for tests. But otherwise ok from my side.

File: returnn/datasets/meta.py (hunk 10 above)
Comment (Stefanwuu) [reply returnn/datasets/meta.py]: There is no unexpected behavior if both parameters are set(but probably redundant), so I don't know whether that would make sense.

Comment (albertz) [reply returnn/datasets/meta.py]: Yea, for me, also not strictly necessary.
(@NeoLegends, can you re-review?)

**Review by NeoLegends — APPROVED** [review]
🚢
